### PR TITLE
Use QSaveFile on file write

### LINKED ISSUE #4514: LMMS file save fail
LMMS occasionally fails to save. I was bumped about this issue on Discord (@DouglasDGI) but I've seen cases on the old forum and also noticed it myself. I saw it once with a **recover.mmp** and I may have mentioned this at the occasion in a thread on GitHub. In the case presented on discord the file was verified to have a size of 0 bytes.

Examples:
[Can't save projects on Mac osx](https://lmms.io/forum/viewtopic.php?f=7&t=28530)
[Trouble with saving projects on Mac OSX](https://lmms.io/forum/viewtopic.php?f=7&t=8446)

When writing a file, we do test for it being larger than 0 bytes [here](https://github.com/LMMS/lmms/blob/e07f9a3b648d8ce168f247217509677ecd38811d/src/core/DataFile.cpp#L307): 

`	if( QFileInfo( outfile.fileName() ).size() > 0 )`

If going by size like this, maybe we could bump the zero to 100 or so, as the minimum file size with header seem to be at least a hundred bytes.

### PR DESCRIPTION
**Pull request by zonkmachine**
Fixes #4514 by replacing QFile with QSaveFile on save action.

### CODE REVIEW

**Review by Spekular — APPROVED** [review]
LGTM. Merge?

**Comment by Spekular** [comment]
Actually though, aren't we doing duplicate work? QSaveFile already uses a temporary file, so we can probably write directly to the final filename now, no?

**Review by Spekular — CHANGES_REQUESTED** [review]
Approved too fast (only looked at diff). Upon closer inspection, it seems like we should remove our homebrewed temp file since QSaveFile already uses the same method.

**Comment by DomClark** [comment]
Prior to this PR, the backup file was the old version of the project file, before being saved. Now, it is identical to the newly-saved version. Is this intended? It doesn't strike me as particularly useful.

**Comment by zonkmachine** [comment]
Right. Looking into it.

**Comment by zonkmachine** [comment]
After discussion on Discord `#dev-only` fullNameTemp is now back. I just dropped the file size check which is now redundant.
Saving with and without backup file works.

**Comment by zonkmachine** [comment]
Now after the QSaveFile commit() is performed we still do a bunch of QFile actions and the commited file could end up either in fullName + ".new" or being deleted if one of the QFile::remove(...)/QFile::rename(...), for whatever reason, fails.

**Comment by zonkmachine** [comment]
Also, if a save operation fail it returns `false`, but it doesn't look like we act on that everytime we save. We do when using `MainWindow::mayChangeProject()` such as in `MainWindow::closeEvent()`.

File: src/core/DataFile.cpp [hunk 1]
```diff
@@ -387,30 +388,29 @@ bool DataFile::writeFile(const QString& filename, bool withResources)
...
-		{
-			// remove old backup file
-			QFile::remove( fullNameBak );
-			// move current file to backup file
-			QFile::rename( fullName, fullNameBak );
-		}
-		// move temporary file to current file
-		QFile::rename( fullNameTemp, fullName );
+		showError(SongEditor::tr("Could not write file"),
+			SongEditor::tr("An unknown error has occured and the file could not be saved."));
```
Comment (zonkmachine) [inline src/core/DataFile.cpp:394]: outfile.errorString() goes in here...

**Comment by zonkmachine** [comment]
I would be jolly happy if someone whould like to take over this issue/PR. @Spekular?
I don't feel confident enough with the coding here and I'm to rusty anyway. Just grab the code.. :) Plz...

File: src/core/DataFile.cpp (hunk 1 above)
Comment (zonkmachine) [reply src/core/DataFile.cpp:407]: ^2 years later... What is that man on about?!

File: src/core/DataFile.cpp [hunk 2]
```diff
@@ -379,12 +380,12 @@ bool DataFile::writeFile(const QString& filename, bool withResources)
 		}
 	}
 
-	QFile outfile (fullNameTemp);
+	QSaveFile outfile (fullNameTemp);
```
Comment (Veratil) [inline src/core/DataFile.cpp:383]:
```suggestion
	QSaveFile outfile(fullNameTemp);
```

File: src/core/DataFile.cpp [hunk 3]
```diff
@@ -405,30 +406,29 @@ bool DataFile::writeFile(const QString& filename, bool withResources)
...
-		}
-		// move temporary file to current file
-		QFile::rename( fullNameTemp, fullName );
+		showError(SongEditor::tr("Could not write file"),
+			SongEditor::tr("An unknown error has occured and the file could not be saved."));
+		return false;
+	}
 
-		return true;
+	if (ConfigManager::inst()->value("app", "disablebackup" ).toInt())
```
Comment (Veratil) [inline src/core/DataFile.cpp:416]:
```suggestion
	if (ConfigManager::inst()->value("app", "disablebackup").toInt())
```

**Review by Veratil — APPROVED** [review]
Two quick style fixes and good to go

**Comment by zonkmachine** [comment]
@Veratil Thanks!

**Pull request merged**